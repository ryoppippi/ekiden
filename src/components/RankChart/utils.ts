import type { ChartData, ChartOptions } from "chart.js";
import type { Article, Ranking } from "./types";

/**
 * ランキングを計算し、1つの配列にまとめる
 */
export function getRanking(articles: Article[]): Ranking[] {
  /** GitHubユーザーごとの記事数をカウント */
  const userCounts: { [user: string]: number } = {};
  articles.forEach((article) => {
    const user = article.githubUser;
    if (userCounts[user]) {
      userCounts[user]++;
    } else {
      userCounts[user] = 1;
    }
  });

  /** 同率のユーザーをグループ化 */
  const groupedUsers: { [count: number]: { users: string[] } } = {};
  Object.entries(userCounts).forEach(([user, count]) => {
    if (!groupedUsers[count]) {
      groupedUsers[count] = { users: [] };
    }
    groupedUsers[count].users.push(user);
  });

  const userRankings: Ranking[] = [];
  let rank = 1;
  Object.keys(groupedUsers)
    .map(Number)
    .sort((a, b) => b - a)
    .forEach((count) => {
      const users = groupedUsers[count].users;
      users.forEach((user) => {
        userRankings.push({ user, articleCount: count, rank });
      });
      rank += users.length;
    });

  return userRankings.filter(({ rank }) => rank <= 10);
}

/**
 * getOptions returns the options for the bar chart.
 */
export function getOptions(data: ChartData<"bar">) {
  return {
    indexAxis: "y",
    // Elements options apply to all of the options unless overridden in a dataset
    // In this case, we are setting the border of each horizontal bar to be 2px wide
    elements: {
      bar: {
        borderWidth: 2,
      },
    },
    responsive: true,
    maintainAspectRatio: false,
    plugins: {
      legend: {
        display: false,
      },
      title: {
        display: false,
      },
    },
    onClick: function (_, items) {
      if (items.length > 0) {
        const item = items[0];
        if (data.labels === null || data.labels![item.index] === null) return;
        const labelText = data.labels![item.index] + "";

        const matched = /^\d+\.\s+(\S+)/.exec(labelText);
        if (matched === null) return;
        const githubUser = matched[1];
        window.location.href = `/ekiden/runners/${githubUser}/`;
      }
    },
    scales: {
      x: {
        ticks: {
          stepSize: 1,
        },
      },
      y: {
        ticks: {
          autoSkip: false,
          padding: 4,
          font: {
            size: 16,
            family: "monospace",
          },
        },
        grid: {
          display: false,
        },
      },
    },
  } as const satisfies ChartOptions<"bar">;
}

/**
 * getData returns the data for the bar chart.
 */
export function getData(ranking: Ranking[]) {
  /** find the longest username */
  const longestUsername = Math.max(...ranking.map(({ user }) => user.length));

  const rankingTick = ranking.map(({ user, rank }) => {
    // FIXME: 左揃えにする方法がわからなかった
    const padding = " ".repeat(longestUsername - user.length);
    return `${rank}. ${user}${padding}`;
  });

  return {
    labels: rankingTick,
    datasets: [
      {
        label: "記事数",
        data: ranking.map(({ articleCount }) => articleCount),
        barThickness: 12,
        backgroundColor: (ctx) => {
          const rank = ranking[ctx.dataIndex].rank;
          switch (rank) {
            case 1:
              return "#FFD700";
            case 2:
              return "#C0C0C0";
            case 3:
              return "#C47222";
            default:
              return "#6cb26a";
          }
        },
        borderRadius: 5,
        borderWidth: 0,
        barPercentage: 0.5,
      },
    ],
  } as const satisfies ChartData<"bar">;
}
