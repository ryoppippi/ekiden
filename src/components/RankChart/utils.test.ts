import * as utils from "./utils";

describe("RankChart utils.ts", () => {
  describe("getRanking", () => {
    it("should start with rank 1", () => {
      const ranking = utils.getRanking([{ githubUser: "alice" }]);
      expect(ranking[0].rank).toBe(1);
    });

    it("should assign same rank for ties and skip next ranks appropriately", () => {
      const articles = [
        { githubUser: "alice" },
        { githubUser: "alice" },
        { githubUser: "alice" },
        { githubUser: "bob" },
        { githubUser: "bob" },
        { githubUser: "bob" },
        { githubUser: "charlie" },
        { githubUser: "charlie" },
        { githubUser: "david" },
      ];

      const ranking = utils.getRanking(articles);

      expect(ranking).toEqual([
        { rank: 1, user: "alice", articleCount: 3 },
        { rank: 1, user: "bob", articleCount: 3 },
        { rank: 3, user: "charlie", articleCount: 2 },
        { rank: 4, user: "david", articleCount: 1 },
      ]);
    });

    it("should have total article count", () => {
      const ranking = utils.getRanking([
        { githubUser: "alice" },
        { githubUser: "alice" },
        { githubUser: "bob" },
      ]);

      expect(ranking.length).toBe(2);
    });
  });

  describe("getTopNRanking", () => {
    const ranking = utils.getRanking([
      { githubUser: "alice" },
      { githubUser: "alice" },
      { githubUser: "alice" },
      { githubUser: "bob" },
      { githubUser: "bob" },
      { githubUser: "bob" },
      { githubUser: "charlie" },
      { githubUser: "charlie" },
      { githubUser: "david" },
    ]);

    it("should get top 1 rank", () => {
      const top1Ranking = utils.getTopNRanking(ranking, 1);
      expect(top1Ranking).toEqual([
        { rank: 1, user: "alice", articleCount: 3 },
        { rank: 1, user: "bob", articleCount: 3 },
      ]);
    });

    it("should get top 2 ranks (including ties)", () => {
      const top2Ranking = utils.getTopNRanking(ranking, 2);
      expect(top2Ranking).toEqual([
        { rank: 1, user: "alice", articleCount: 3 },
        { rank: 1, user: "bob", articleCount: 3 },
      ]);
    });

    it("should get top 3 ranks", () => {
      const top3Ranking = utils.getTopNRanking(ranking, 3);
      expect(top3Ranking).toEqual([
        { rank: 1, user: "alice", articleCount: 3 },
        { rank: 1, user: "bob", articleCount: 3 },
        { rank: 3, user: "charlie", articleCount: 2 },
      ]);
    });
  });

  describe("getData", () => {
    const ranking = utils.getRanking([
      { githubUser: "alice" },
      { githubUser: "alice" },
      { githubUser: "alice" },
      { githubUser: "bob" },
      { githubUser: "bob" },
      { githubUser: "bob" },
      { githubUser: "charlie" },
      { githubUser: "charlie" },
      { githubUser: "david" },
    ]);
    const top3Ranking = utils.getTopNRanking(ranking, 3);

    it("should match getData snapshot", () => {
      const data = utils.getData(top3Ranking);
      expect(data).toMatchInlineSnapshot(`
        {
          "datasets": [
            {
              "backgroundColor": [Function],
              "barPercentage": 0.5,
              "barThickness": 12,
              "borderRadius": 5,
              "borderWidth": 0,
              "data": [
                3,
                3,
                2,
              ],
              "label": "記事数",
            },
          ],
          "labels": [
            "1. alice  ",
            "1. bob    ",
            "3. charlie",
          ],
        }
      `);
    });
  });

  describe("getOptions", () => {
    const ranking = utils.getRanking([]);

    it("should match getOptions snapshot", () => {
      const data = utils.getData(ranking);
      const options = utils.getOptions(data);
      expect(options).toMatchInlineSnapshot(`
        {
          "elements": {
            "bar": {
              "borderWidth": 2,
            },
          },
          "indexAxis": "y",
          "maintainAspectRatio": false,
          "onClick": [Function],
          "plugins": {
            "legend": {
              "display": false,
            },
            "title": {
              "display": false,
            },
          },
          "responsive": true,
          "scales": {
            "x": {
              "ticks": {
                "stepSize": 1,
              },
            },
            "y": {
              "grid": {
                "display": false,
              },
              "ticks": {
                "autoSkip": false,
                "font": {
                  "family": "monospace",
                  "size": 16,
                },
                "padding": 4,
              },
            },
          },
        }
      `);
    });
  });
});
