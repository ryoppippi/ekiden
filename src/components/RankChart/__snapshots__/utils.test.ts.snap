// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RankChart utils.ts > getRanking 1`] = `
[
  {
    "articleCount": 5,
    "rank": 1,
    "user": "staticWagomU",
  },
  {
    "articleCount": 5,
    "rank": 1,
    "user": "uga-rosa",
  },
  {
    "articleCount": 4,
    "rank": 3,
    "user": "monaqa",
  },
  {
    "articleCount": 3,
    "rank": 4,
    "user": "atusy",
  },
  {
    "articleCount": 3,
    "rank": 4,
    "user": "ryoppippi",
  },
  {
    "articleCount": 3,
    "rank": 4,
    "user": "kyoh86",
  },
  {
    "articleCount": 2,
    "rank": 7,
    "user": "thinca",
  },
  {
    "articleCount": 2,
    "rank": 7,
    "user": "matoruru",
  },
  {
    "articleCount": 2,
    "rank": 7,
    "user": "ArcCosine",
  },
  {
    "articleCount": 2,
    "rank": 7,
    "user": "eihigh",
  },
  {
    "articleCount": 2,
    "rank": 7,
    "user": "skanehira",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "tani",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "kuuote",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "Liquid-system",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "fuzmare",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "tadashi-aikawa",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "get-me-power",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "hokorobi",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "willelz",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "kamecha",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "ryota2357",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "ikorihn",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "mityu",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "sontixyou",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "yukimemi",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "uhooi",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "kawarimidoll",
  },
  {
    "articleCount": 1,
    "rank": 12,
    "user": "futsuuu",
  },
]
`;
